server/channels/jobs/batch_report_worker.go

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package jobs
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/channels/store"
+	"github.com/pkg/errors"
+)
+
+type BatchReportWorkerAppIFace interface {
+	SaveReportChunk(format string, prefix string, count int, reportData []model.ReportableObject) *model.AppError
+	CompileReportChunks(format string, prefix string, numberOfChunks int, headers []string) *model.AppError
+	SendReportToUser(rctx request.CTX, userID string, jobId string, format string) *model.AppError
+	CleanupReportChunks(format string, prefix string, numberOfChunks int) *model.AppError
+}
+
+type BatchReportWorker[T BatchReportWorkerAppIFace] struct {
+	BatchWorker
+	app          T
+	reportFormat string
+	headers      []string
+	getData      func(jobData model.StringMap, app T) ([]model.ReportableObject, model.StringMap, bool, error)
+}
+
+func MakeBatchReportWorker[T BatchReportWorkerAppIFace](
+	jobServer *JobServer,
+	store store.Store,
+	app T,
+	timeBetweenBatches time.Duration,
+	reportFormat string,
+	headers []string,
+	getData func(jobData model.StringMap, app T) ([]model.ReportableObject, model.StringMap, bool, error),
+) model.Worker {
+	worker := &BatchReportWorker[T]{
+		app:          app,
+		reportFormat: reportFormat,
+		headers:      headers,
+		getData:      getData,
+	}
+	worker.BatchWorker = BatchWorker{
```

Comment:
nit: Should we inline setting `BatchWorker` above vs. assigning it discretely?

Reply:
The `BatchWorker` module requires a `doBatch` parameter to call since it handles the actual batching calls. This is provided by the `BatchReportWorker` struct, so we have to attach it after.

I can be a bit clearer with that and set it separately instead of the whole thing.